### DISCUSSION TITLE
Remove unused imports recipe is removing an used class

### DISCUSSION THREAD

**Discussion by sergiogp98** [openrewrite/rewrite#3903] [General] [answered]
Hi,

I am running a single recipe (https://docs.openrewrite.org/recipes/java/removeunusedimports) and is removing some imports of a couple of Java classes which they are already in use, hence it is failing when it tries to compile.

My question: which criteria is used to know if an import is used? I don't understand the explanation in your official documentation:

> Remove imports for types that are not referenced. As a precaution against incorrect changes, no imports will be removed from any source where unknown types are referenced.

For example, one of the imports which is removed is this one: _org.springframework.security.web.SecurityFilterChain_

Thanks.

**Comment by timtebeek** [comment, marked as answer]
Hi @sergiogp98  ; did you already look at the implementation of the recipe?
https://github.com/openrewrite/rewrite/blob/c625f30ca32f8af7a196dd3b00b720919db22dcf/rewrite-java/src/main/java/org/openrewrite/java/RemoveUnusedImports.java#L55-L57

Roughly speaking we look at the types in use on Lossless Semantic Tree elements in the compilation unit; if such types are missing, then we are at risk of misidentifying a used type as unused. If you have any suggestions for improvement we're more than open to those!

**Reply by sergiogp98** [reply]
Hi @timtebeek, thanks for the answer. However, I am not very sure how LST works which makes it difficult for me to understand the criteria behind that. 
I had to use the whole package name instead of the class name. Instead of having this:

```java
import org.springframework.security.web.SecurityFilterChain;
@Bean    
public SecurityFilterChain webSecurityCustomizer(HttpSecurity http) throws Exception {
      
}
```

I have to refactor as this:

```java
@Bean
public org.springframework.security.web.SecurityFilterChain webSecurityCustomizer(HttpSecurity http) throws Exception {
  
}
```

Because line _import org.springframework.security.web.SecurityFilterChain_ is automatically removed by the recipe. Is there a way to prevent that change? I would like the recipe to remove only the unused imports.

**Reply by timtebeek** [reply]
Hi @sergiogp98 ; sounds like a potential bug with the recipe if that particular import is not retained; would you mind logging that as an issue over on https://github.com/openrewrite/rewrite-spring/issues ?
Indeed we want to keep that import around, but it's not yet clear what causes it to be removed.

**Reply by timtebeek** [reply]
Might match up with this bug report we got last week:
- https://github.com/openrewrite/rewrite/issues/3909

**Reply by timtebeek** [reply]
This has since likely been fixed in 
- https://github.com/openrewrite/rewrite/pull/3914

Thanks for pointing it out!

**Reply by sergiogp98** [reply]
I have tested and now is not removing that import. Thanks 😄
